datalad_metalad/add.py

```diff
@@ -0,0 +1,449 @@
+      from the file. For example, add metadata stored in the file
+      "metadata-123.json" to the metadata model instance in the current
+      directory and overwrite metadata values with the values stored in
+      stored in "extra-info.json".
+
+      $ datalad meta-add --metadata-store /metadata/dataset_0 \
+       metadata-123.json @extra-info.json
+
+    """
+
+    required_keys = (
+        "extractor_name",
+        "extractor_version",
+        "extraction_parameter",
+        "extraction_time",
+        "agent_name",
+        "agent_email",
+        "dataset_id",
+        "dataset_version",
+        "extracted_metadata")
+
+    optional_keys = (
+        "intra_dataset_path",)
+
+    required_additional_keys = (
+        "root_dataset_id",
+        "root_dataset_version",
+        "inter_dataset_path")
+
+    required_keys_lines = "\n".join(map(repr, required_keys))
+    required_additional_keys_lines = "\n".join(
+        map(repr, required_additional_keys))
+
+    _params_ = dict(
+            metadata=Parameter(
+            args=("metadata",),
+            metavar="METADATA",
+            doc=f"""Path of a file that contains the metadata that
+            should be added to the metadata model instance (the
+            metadata must be provided as a JSON-serialized metadata
+            dictionary).
+            
+            If the path is "-", metadata is read from standard input.
+            
+            The dictionary must contain the following keys:
+            
+            {required_keys_lines}
+            
+            If the metadata is associated with a file, the following key
+            indicates the file path:
+            
+            'intra_dataset_path'
+            
+            It may in addition contain either all or none of the
+            following keys (they are used to add metadata element
+            as a sub-dataset element, i.e. perform aggregation):
+            
+            {required_additional_keys_lines}            
+            """,
+            constraints=EnsureStr() | EnsureNone()),
+        metadata_store=Parameter(
+            args=("-s", "--metadata-store"),
+            doc="""Directory in which the metadata model instance is
+            stored. If no directory name is provided, the current working
+            directory is used.""",
+            constraints=EnsureStr() | EnsureNone()),
+        additionalvalues=Parameter(
+            args=("additionalvalues",),
+            metavar="ADDITIONAL_VALUES",
+            doc="""A string that contains a JSON serialized dictionary of
+            key value-pairs. These key values-pairs are used in addition to
+            the key value pairs in the metadata dictionary to describe
+            the metadata that should be added. If an additional key is
+            already present in the metadata, it will override the value
+            from metadata. In this case a warning is issued.""",
```

Comment:
That partially answers my question from above. Maybe worth clarifying whether this mechanism works for arbitrarily deeply nested objects, or just for top-level keys.

Reply:
Good point. We should restrict it to some keys. For example, you would not want to overwrite metadata-content.

Additional keys are mainly intended to allow aggregation of existing metadata and possibly to amend something like a provenance tracking list.